File: scripts/mgcl_tracker.py
import os
import sys
import argparse
import textwrap
import re
import logging
from pathlib import Path
import pandas as pd


def error_message(message):
    print("mgcl_tracker.py: error:", message)
    sys.exit(1)


class Tracker:
    def __init__(self, start_dir, _range, nums, exts):
        self.start_dir = start_dir

        self.lower = int(_range[0]) if _range is not None else None
        self.upper = int(_range[1]) if _range is not None else None

        self.exts = tuple(exts)

        # if nums exists, accounted_for will be assigned to it, otherwise I generate it from the
        # range values
        self.accounted_for = {
            i: False for i in range(self.lower, self.upper + 1)} if nums is None else nums

        self.missing = dict()

        logging.basicConfig(
            level=logging.NOTSET,
            filename=os.path.join(start_dir, "tracker.log"),
            filemode="w",
            format="%(levelname)s - %(message)s",
        )

    def extract_number(self, file):
        regex = re.compile(r"(?:MGCL_)(\d*)(?:.*)")
        return regex.findall(file)

    def check_file(self, file):
        nums = self.extract_number(file)

        if not nums or len(nums) > 1:
            logging.error(
                "Unexpected file naming scheme @ {}: expected single number, recieved: {}".format(
                    file, nums
                )
            )
            return

        try:
            num = int(nums[0])

            if num < self.lower or num > self.upper:
                logging.debug("{} in {} not in range...".format(num, file))
                return
            elif num not in self.accounted_for:
                logging.error(
                    "unexpected error: number {} extracted from {} is not in dictionary but is somehow in range...".format(
                        num, file
                    )
                )
                return
            elif not self.accounted_for[num]:
                logging.debug("{} in {} is in range...".format(num, file))
                self.accounted_for[num] = True
        except ValueError:
            logging.error(
                "Could not parse number in filename: {}".format(nums[0]))
        except Exception as e:
            logging.error("unknwon error occurred: {}".format(e))

    def write_missing(self):
        print(
            "Logging numbers that could not be located (this may take time)...", end=""
        )

        with open(
            os.path.join(self.start_dir, "tracker_missing_numbers.txt"), "w"
        ) as f:
            for num, _ in self.missing.items():
                f.write("{}\n".format(num))

        print(" done!")

    def write_found(self):
        print(
            "Logging numbers that were located (this may take time)...",
            end="",
        )

        with open(os.path.join(self.start_dir, "tracker_found_numbers.txt"), "w") as f:
            for num, _ in self.accounted_for.items():
                f.write("{}\n".format(num))

        print(" done!\n")

    def write_out(self):
        self.write_missing()
        self.write_found()

        print(
            len(self.missing),
            "missing numbers total, written to",
            os.path.abspath(
                os.path.join(self.start_dir, "tracker_missing_numbers.txt")
            ),
            end="\n\n",
        )

        print(
            len(self.accounted_for),
            "found numbers total, written to",
            os.path.abspath(os.path.join(
                self.start_dir, "tracker_found_numbers.txt")),
        )

    def run(self):
        print("Starting os walk...", end="")
        for p, d, f in os.walk(self.start_dir):
            for file in f:
                if file.endswith(self.exts):
                    self.check_file(file)
        print(" done!\n")

        print("All files handled...\n")

        self.missing = dict(
            filter(lambda el: el[1] == False, self.accounted_for.items())
        )

        self.accounted_for = dict(
            filter(lambda el: el[1] == True, self.accounted_for.items())
        )

        self.write_out()


def extract_number(file):
    regex = re.compile(r"(?:MGCL_)(\d*)(?:.*)")

    nums = regex.findall(file)

    if not nums or len(nums) > 1:
        error_message(
            "unexpected cataglogNumber naming scheme @ {}: expected single number, recieved: {}".format(
                file, nums
            )
        )

    else:
        try:
            return int(nums[0])
        except ValueError:
            error_message(
                "Could not parse number in cataglogNumber: {}".format(nums[0]))
        except Exception:
            error_message(
                "Unknown error occurred when attempting to parse {}".format(nums[0]))


def parse_csv(csv_file):
    path_obj = Path(csv_file)

    ext = path_obj.suffix

    if ext is None:
        error_message("input_file missing extension")

    ext = ext.replace(".", "")

    if ext.lower() != "csv" and ext.lower() != "xlsx":
        error_message(
            "input_file extension {} must match either csv or xlsx (ignoring case)".format(
                ext
            )
        )

    raw_data = None

    if ext.lower() == "csv":
        raw_data = pd.read_csv(csv_file)
    elif ext.lower() == "xlsx":
        raw_data = pd.read_excel(csv_file)
    else:
        error_message(
            "Invalid file provided: expected .csv or .xlsx, recieved {}".format(ext.lower()))

    try:
        # this is a long complicated line so I will break it down
        # I create lambda to take an array of MGCL numbers and convert it to a dict { mgcl_number : bool }
        # I zip the raw data and an array of False values together and convert it to a dict
        # I call the lambda using the dataframe column 'catalogNumber' and convert to an array, and call extract number on
        # each element to grab JUST the number from the MGCL_#### pattern
        # I return the result of the invoked lambda function
        return (lambda raw_data: dict(zip(raw_data, [False for _ in range(len(raw_data))])))(
            [extract_number(i) for i in raw_data['catalogNumber'].tolist()]
        )
    except Exception:
        error_message(
            'Invalid formatting in .csv/.xlsx file: expected column \'catalogNumber\'')


def cli():
    parser = argparse.ArgumentParser(
        description="Track the used MGCL numbers in the filesystem",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=textwrap.dedent(
            """\
          Example Runs:
            python3 ./mgcl_tracker.py --start_dir /fake/path --lower 1234567 --upper 2000000 \n
            python3 ./mgcl_tracker.py --start_dir /fake/path --lower 1234567 --upper 2000000 --exts png jpg
         """
        ),
    )

    parser.add_argument(
        "-d",
        "--start_dir",
        required=True,
        type=str,
        help="path to the starting directory",
    )

    parser.add_argument(
        "-e",
        "--exts",
        required=False,
        nargs="+",
        default=["png", "jpg", "jpeg"],
        help="The upper bound MGCL number to include in count (default is png jpg jpeg)",
    )

    group = parser.add_mutually_exclusive_group(required=True)

    group.add_argument(
        "-r",
        "--range",
        nargs=2,
        help="The lower and upper bound MGCL numbers"
    )

    group.add_argument(
        '-f',
        '--file',
        help="A CSV of MGCL Numbers to search for"
    )

    args = parser.parse_args()

    start_dir = args.start_dir
    csv_file = args.file
    _range = args.range
    exts = args.exts

    nums = parse_csv(csv_file) if csv_file is not None else None

    print(
        "**WARNING: This script indexes all files recursively from the start. As a result, this can take an exceptional amount of time for large targets.**\n"
    )

    print("Program starting...\n")

    tracker = Tracker(start_dir, _range, nums, exts)
    tracker.run()

    print("\nAll computations completed...")


if __name__ == "__main__":
    cli()

File: scripts/mgcl_tracker.jl
using ArgParse
using CSV
using DataFrames
using Logging

# THIS FILE IS NOT INTENDED FOR USE! 
# I am trying to learn Julia by converting my latest script.

# there are no classes in Julia... so this will be interesting


# TODO: figure me out plz
# io = open("log.txt", "w+")
# logger = SimpleLogger(io)

# this is all for the most part the same as I would define my CLI args in python.
# I don't particularly love the syntax for the arg tables, but it isn't that bad.
function parse_commandline()
    s = ArgParseSettings()

    add_arg_group!(s, "Mutually exclusive options", exclusive = true, required = true)
    @add_arg_table! s begin
        "--range", "-r"
        nargs = 2
        help = "The lower and upper bound MGCL numbers, from LOWER to UPPER (inclusive)"
        metavar = ["LOWER", "UPPER"]
        arg_type = Int

        "--file", "-f"
        help = "A CSV/XLSX of MGCL Numbers to search for"
    end

    add_arg_group!(s, "Required arguments", required = true)
    @add_arg_table s begin
        "--start_dir", "-d"
        help = "path to the starting directory"
        required = true

        "--exts", "-e"
        help = "The upper bound MGCL number to include in count (default is png jpg jpeg)"
        default = ["png", "jpg", "jpeg"]
        required = false
    end

    return parse_args(s)
end


function extract_number(filename)
    pattern = r"(?:MGCL_)(\d*)(?:.*)"

    numbers = map(eachmatch(pattern, filename)) do m
        parse(Int, m[1])
    end

    if length(numbers) > 1
        # handle me
    end

    # oh no.... this took me an hour to figure out that Julia is 
    # indexed starting at ONE :( By far my least favorite Data Science
    # thing. The errors Julia produces are kinda eh to be honest.
    # See rust for amazing error output, sometimes I swear rust has
    # ai in their compiler.
    return numbers[1]
end


function parse_file(_file)
    # normally I would not do this in three separate patterns,
    # however I could not find a concise way of extracting the suffix of a
    # file _safely_. I python I just use Path objects.
    csv_pattern = r".*\.csv$"
    xlsx_pattern = r".*\.xlsx$"
    both_pattern = r".*(?:\.csv$|\.xlsx$)"

    # null would have been quicker to type lol
    raw_data = nothing

    # so naturally if you are going to have an isnothing in your std lib then
    # issomething should also be there. It's not a big deal, I'm just being 
    # a brat.
    if isnothing(match(both_pattern, _file))
        # TODO throw and error

    elseif isnothing(match(csv_pattern, _file))
        # handle me
        # raw_data = read_excel(_file)
    else
        # in python I would use pandas for almost everything and everthing like this,
        # however the Pandas package for Julia is actually a wrapper around the package.
        # So to try and separate this file I decided not to use it.
        raw_data = DataFrame(CSV.File(_file, header = 1))
    end

    accounted_for = Dict{Int,Bool}()

    for r in eachrow(raw_data)
        accounted_for[extract_number(r.catalogNumber)] = false
    end

    return accounted_for
end


function parse_range(range)
    lower = range[1]
    upper = range[2]

    if upper < lower
        # TODO throw an error
    end

    accounted_for = Dict{Int,Bool}()

    for i = lower:upper
        accounted_for[i] = false
    end

    return accounted_for
end


function walk(accounted_for, start_dir, exts, range)
    ext_string = join(exts, "|")
    pattern = Regex(".*($ext_string)\$")



    print("Starting os walk...")

    # TODO: this needs error handling
    for (root, dirs, files) in walkdir(start_dir)
        for file in files
            if !isnothing(match(pattern, file))
                num = extract_number(file)

                if !isnothing(range)
                    lower = range[1]
                    upper = range[2]

                    if num < lower || num > upper
                        # TODO handle me
                        @debug "Number $num in $file is out of range"
                    elseif !haskey(accounted_for, num)
                        # TODO unexpected error, this should not happen
                        @warn "Number $num in $file is in range but was somehow excluded (is it an integer?)"
                    elseif accounted_for[num] == false
                        accounted_for[num] = true
                        @debug "Found number $num in $file"
                    end

                elseif haskey(accounted_for, num) && accounted_for[num] == false
                    accounted_for[num] = true
                    @debug "Found number $num in $file"
                end
            end

        end
    end

    println(" done!")

    return accounted_for

end

function main()
    parsed_args = parse_commandline()

    exts = parsed_args["exts"]
    range = parsed_args["range"]
    start_dir = parsed_args["start_dir"]
    _file = parsed_args["file"]

    accounted_for = nothing

    if !isnothing(_file)
        accounted_for = parse_file(_file)

    else
        accounted_for = parse_range(range)
    end

    updated = walk(accounted_for, start_dir, exts, range)

    finds = filter(entry -> entry.second == true, updated)

    print(finds)
end

main()

# Notes
#
# I'm making this to learn the language, I'll make some notes and comments both 
# throughout the code and down below.
#
####################################################################################
#
# Dictionaries! -> https://docs.julialang.org/en/v1/base/collections/#Dictionaries
#
# Basically the same as python. Curly bracket initialization has been deprecated,
# so instantiation is along the lines of:
#
# myDict = Dict() --> creates Dict{Any,Any}()
# myDict["firstkey"] = "firstvalue" --> creates Dict{Any,Any} with 1 entry: "firstkey" => "firstvalue"
#
# You can check existance of keys AND key-value pairs which is cool:
#
# haskey(myDict, "Z") --> false
# in(("firstkey" => "firstvalue"), myDict) --> true
#
# see more --> https://en.wikibooks.org/wiki/Introducing_Julia/Dictionaries_and_sets
#
####################################################################################
#
# TODO more notes
#
####################################################################################


# https://discourse.julialang.org/t/why-is-os-walk-regex-so-much-slower-than-glob/43906
#
# Apparently there were issues with Julias os walk implementation (see above link). I'll 
# be sure to create two variants to test whether or not the issues were actually fixed.
#
##########################################
#
# OPTION 1 - glob (had no issues)
#
# files = glob("MATCH_GLOB", ".")
# for file in files 
#     if occursin("MisleadingDirectoryName", file)  && continue
#     #do input
#     end
# end
#
##########################################
#
# OPTION 2 - walkdir (had issues)
#
# for (root, dirs, files) in walkdir(".")
#     #This is how I handled not going into directories with false information that I didn't want to parse later
#      deleteat!(dirs,findall(x->"MisleadingDirectoryName",dirs))
#      for file in files
#            if match(MATCH_REGEX, file).match
#                 #do input
#            end
#       end
# end
#
##########################################